singleFrameProcCoreTasks/charimg.rst

```diff
@@ -0,0 +1,304 @@
+
+See `lsstDebug.info`_ for more on the debugging framework.
+
+.. _`lsstDebug.info`: https://lsst-web.ncsa.illinois.edu/doxygen/x_masterDoxyDoc/classlsst_debug_1_1_info.html 
+
+
+
+Examples
+========
+
+.. This example is not working in the current stack (see https://jira.lsstcorp.org/browse/DM-9142)  --- 2/9/2017
+   
+
+Note: running this example currently requires that over and above the DM Stack installation, `afwdata`_ is installed and set up (via the EUPS `setup <https://dev.lsstcorp.org/trac/wiki/EupsTutorial>`_ command).
+.. This is a general link to the EUPS tutorial, but setup is explained in there
+.. _`afwdata`: https://github.com/lsst/afwdata
+
+This example script is `calibrateTask.py` (which calls :doc:`CharacterizeImageTask <apiUsage_charimg>`) before calling :doc:`CalibrateTask <calibimg>` in the `$PIPE_TASKS/examples` directory, and can be run from the command line as, e.g.:
+
+
+.. code-block:: python
+  
+     python examples/calibrateTask.py -display
+
+Where the `-display` flag tells the script to bring up the display tool to show the image files after each step.
+     
+The first thing the example does is import the task (there are some other standard imports as well that are not extracted out here):
+
+.. code-block:: python
+		
+    from lsst.pipe.tasks.characterizeImage import CharacterizeImageTask
+
+The script next processes the data. This occurs in two steps:
+
+- Characterize the image: measure bright sources, fit a background and PSF, and repairs cosmic rays
+
+.. code-block:: python
+		
+     exposure = loadData()
+     exposureIdInfo = ExposureIdInfo(expId=1, expBits=5)
+ 
+     # characterize the exposure to repair cosmic rays and fit a PSF model
+     # display now because CalibrateTask modifies the exposure in place
+     charRes = charImageTask.characterize(exposure=exposure, exposureIdInfo=exposureIdInfo)
+     if display:
+         displayFunc(charRes.exposure, charRes.sourceCat, frame=1)
+
+- Calibrate the exposure: measure faint sources, fit an improved WCS and photometric zero-point
+		
+.. code-block:: python
+
+   
+    # calibrate the exposure
+    calRes = calibrateTask.calibrate(exposure=charRes.exposure, exposureIdInfo=exposureIdInfo)
+    if display:
+        displayFunc(calRes.exposure, calRes.sourceCat, frame=2)
+
+To round out this minimal description, the `displayFunc` that is called above in the blocks is defined as so:
+
+.. code-block:: python
+		
+ def displayFunc(exposure, sourceCat, frame):
+    display = afwDisplay.getDisplay(frame)
+    display.mtv(exposure)
+
+    with display.Buffering():
+        for s in sourceCat:
+            xy = s.getCentroid()
+            display.dot('+', *xy, ctype=afwDisplay.CYAN if s.get("flags_negative") else afwDisplay.GREEN)
+	
+     
+
+Algorithm details
+=================
+
```

Comment:
Clearly, as a group, we should discuss what we're hoping to get out of this section in more detail.

Somewhere in the task documentation, I think we need a brief functional description of what's happening. It can assume the reader knows the material in the introduction, and that they understand the ultimate aim of the process (we do not need to educate them about why they want a good background model, for example). I'm thinking of something like the following, perhaps:

> The PSF is iteratively arrived at by repeatedly interpolating over cosmic rays (delegated to the `repair` retargetable subtask),  estimating and subtracting the background (the `background` subtask), detecting (`detection`), deblending (`deblend`) and measuring (`measurement`) sources, and using those sources to estimate the PSF (`measurePsf`). This is repeated `psfIterations` times, gradually refining the PSF model. After the ultimate PSF has been so derived, it is used in final `repair` and `measurement` steps which produce the source catalogue returned to the caller.

(NB scanning the code I'm not sure the above is strictly accurate, but I think it gives the gist of what's appropriate.)

I'm actually not sure whether it's necessary to refer to specific code objects or not. I think it's probably useful, since the reader likely cares that all these algorithms are retargetable, but you could likely persuade me either way.

I think a brief paragraph like the above feels at home in “Algorithmic Details”, but you could probably convince me that it belongs in the introduction or something instead.

In addition, depending on the contents of the task, it _may_ be appropriate to provide one or both of:

- References to published papers, technical notes, etc, describing the algorithm.
- Where external descriptions of the mathematics do not exist, a brief mathematical description of what's happening.

I don't think these are always necessary, particularly for high-level tasks which are effectively aggregations of others (like `CharacterizeImageTask`). Where they are necessary, they form the core of the “details” section.